src/API/Instrumentation/AutoInstrumentation/ExtensionHookManager.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\API\Instrumentation\AutoInstrumentation;
+
+use function assert;
+use Closure;
+use function extension_loaded;
+use Nevay\SPI\ServiceProviderDependency\ExtensionDependency;
+use OpenTelemetry\Context\Context;
+use OpenTelemetry\Context\ContextInterface;
+use OpenTelemetry\Context\ContextKeyInterface;
+use ReflectionFunction;
+
+/** @phan-file-suppress PhanUndeclaredClassAttribute */
+
+#[ExtensionDependency('opentelemetry', '^1.0')]
+final class ExtensionHookManager implements HookManager
+{
+
+    private readonly ContextKeyInterface $contextKey;
+
+    public function __construct()
+    {
+        $this->contextKey = Context::createKey(self::class);
+    }
+
+    /**
+     * @phan-suppress PhanUndeclaredFunction
+     */
+    public function hook(?string $class, string $function, ?Closure $preHook = null, ?Closure $postHook = null): void
+    {
+        assert(extension_loaded('opentelemetry'));
+
+        /** @noinspection PhpFullyQualifiedNameUsageInspection */
+        \OpenTelemetry\Instrumentation\hook($class, $function, $this->bindHookScope($preHook), $this->bindHookScope($postHook));
+    }
+
+    public function enable(ContextInterface $context): ContextInterface
+    {
+        return $context->with($this->contextKey, true);
+    }
+
+    public function disable(ContextInterface $context): ContextInterface
+    {
+        return $context->with($this->contextKey, null);
+    }
+
+    private function bindHookScope(?Closure $closure): ?Closure
+    {
+        if (!$closure) {
+            return null;
+        }
+
+        $contextKey = $this->contextKey;
+        $reflection = new ReflectionFunction($closure);
+
+        // TODO Add an option flag to ext-opentelemetry `hook` that configures whether return values should be used?
+        if (!$reflection->getReturnType() || (string) $reflection->getReturnType() === 'void') {
+            return static function (mixed ...$args) use ($closure, $contextKey): void {
```

Comment:
question from SIG meeting: do we understand the difference between these two variants? does it affect a method's return type in unexpected ways?

Reply:
I think this is resolved (and covered by tests) now.